File: src/windows/Maple.WinUI/MapleUI/Organisms/MuiSelectivePasteModal.cs
using System;
using System.Collections.Generic;
using System.Linq;
using Microsoft.UI.Xaml;
using Microsoft.UI.Xaml.Controls;
using Maple.UI.Atoms;

namespace Maple.UI
{
    /// <summary>One toggleable adjustment group in a Selective Paste run.</summary>
    public sealed record MuiSelectivePasteGroup(string Id, string Label);

    /// <summary>
    /// Maple.UI Selective Paste modal organism (unified-component-catalog.md
    /// §4.4, "Selective Paste" row: "Per-group apply toggles", built from
    /// Checkbox, Text, Button) — one <see cref="MuiCheckbox"/> per tool
    /// group (Light, Color, Effects, …), each labeled with a
    /// <see cref="MuiText"/>, gating which groups a Paste actually
    /// applies.
    /// </summary>
    public sealed class MuiSelectivePasteModal : ContentControl
    {
        public static readonly DependencyProperty IsOpenProperty =
            DependencyProperty.Register(nameof(IsOpen), typeof(bool), typeof(MuiSelectivePasteModal),
                new PropertyMetadata(false, (d, e) => ((MuiSelectivePasteModal)d)._shell.IsOpen = (bool)e.NewValue));

        public static readonly DependencyProperty ContainedProperty =
            DependencyProperty.Register(nameof(Contained), typeof(bool), typeof(MuiSelectivePasteModal),
                new PropertyMetadata(false, (d, e) => ((MuiSelectivePasteModal)d)._shell.Contained = (bool)e.NewValue));

        public static readonly DependencyProperty GroupsProperty =
            DependencyProperty.Register(nameof(Groups), typeof(IReadOnlyList<MuiSelectivePasteGroup>), typeof(MuiSelectivePasteModal),
                new PropertyMetadata(null, (d, _) => ((MuiSelectivePasteModal)d).Rebuild()));

        public static readonly DependencyProperty SelectedGroupIdsProperty =
            DependencyProperty.Register(nameof(SelectedGroupIds), typeof(IReadOnlyList<string>), typeof(MuiSelectivePasteModal),
                new PropertyMetadata(null, (d, _) => ((MuiSelectivePasteModal)d).Rebuild()));

        public bool IsOpen { get => (bool)GetValue(IsOpenProperty); set => SetValue(IsOpenProperty, value); }
        public bool Contained { get => (bool)GetValue(ContainedProperty); set => SetValue(ContainedProperty, value); }

        public IReadOnlyList<MuiSelectivePasteGroup>? Groups
        {
            get => (IReadOnlyList<MuiSelectivePasteGroup>?)GetValue(GroupsProperty);
            set => SetValue(GroupsProperty, value);
        }

        public IReadOnlyList<string>? SelectedGroupIds
        {
            get => (IReadOnlyList<string>?)GetValue(SelectedGroupIdsProperty);
            set => SetValue(SelectedGroupIdsProperty, value);
        }

        public event EventHandler? Dismissed;
        public event EventHandler<IReadOnlyList<string>>? PasteRequested;

        private readonly MuiOverlayShell _shell = new() { Size = MuiOverlayShellSize.Sm, AriaLabel = "Selective Paste" };
        private readonly StackPanel _checks = new() { Orientation = Orientation.Vertical, Spacing = 8 };
        private readonly MuiButton _cancel = new() { Variant = MuiButtonVariant.Ghost, Label = "Cancel" };
        private readonly MuiButton _paste = new() { Variant = MuiButtonVariant.Primary, Label = "Paste" };

        public MuiSelectivePasteModal()
        {
            var footer = new StackPanel { Orientation = Orientation.Horizontal, Spacing = 8, HorizontalAlignment = HorizontalAlignment.Right };
            footer.Children.Add(_cancel);
            footer.Children.Add(_paste);

            _shell.Header = new MuiText { Text = "Selective Paste", Variant = MuiTextVariant.SheetTitle };
            _shell.Body = _checks;
            _shell.Footer = footer;
            Content = _shell;
            HorizontalContentAlignment = HorizontalAlignment.Stretch;
            VerticalContentAlignment = VerticalAlignment.Stretch;

            _shell.Dismissed += (_, _) => { IsOpen = false; Dismissed?.Invoke(this, EventArgs.Empty); };
            _cancel.Click += (_, _) => { IsOpen = false; Dismissed?.Invoke(this, EventArgs.Empty); };
            _paste.Click += (_, _) => PasteRequested?.Invoke(this, SelectedGroupIds ?? Array.Empty<string>());

            Rebuild();
        }

        private void Rebuild()
        {
            var selected = SelectedGroupIds is null ? new HashSet<string>() : new HashSet<string>(SelectedGroupIds);
            _checks.Children.Clear();
            foreach (var group in Groups ?? Array.Empty<MuiSelectivePasteGroup>())
            {
                var checkbox = new MuiCheckbox { Label = group.Label, CheckedState = selected.Contains(group.Id) };
                var groupId = group.Id;
                checkbox.Checked += (_, _) => Toggle(groupId, true);
                checkbox.Unchecked += (_, _) => Toggle(groupId, false);
                _checks.Children.Add(checkbox);
            }
        }

        private void Toggle(string groupId, bool selected)
        {
            var current = SelectedGroupIds is null ? new HashSet<string>() : new HashSet<string>(SelectedGroupIds);
            if (selected) current.Add(groupId); else current.Remove(groupId);
            SelectedGroupIds = current.ToList();
        }
    }
}

File: src/apple/Packages/MapleUI/Sources/MapleUI/OrganismsB/MuiSelectivePasteModal.swift
// MuiSelectivePasteModal.swift — Maple UI Organisms · Modals (unified-
// component-catalog.md §4.4). Per-group apply toggles before pasting
// settings onto other photos, built on Overlay Shell from Checkbox
// (repeated per group), Text, and Button.

import SwiftUI

public struct MuiSelectivePasteGroup: Identifiable, Sendable {
    public let id: String
    public let label: String
    public let description: String?
    public let enabled: Bool

    public init(id: String, label: String, description: String? = nil, enabled: Bool) {
        self.id = id
        self.label = label
        self.description = description
        self.enabled = enabled
    }
}

public struct MuiSelectivePasteModal: View {
    public let isPresented: Bool
    public let contained: Bool
    @Binding public var groups: [MuiSelectivePasteGroup]
    public let pasteConfirmed: (([String]) -> Void)?
    public let dismissed: (() -> Void)?

    public init(
        isPresented: Bool,
        contained: Bool = false,
        groups: Binding<[MuiSelectivePasteGroup]>,
        pasteConfirmed: (([String]) -> Void)? = nil,
        dismissed: (() -> Void)? = nil
    ) {
        self.isPresented = isPresented
        self.contained = contained
        self._groups = groups
        self.pasteConfirmed = pasteConfirmed
        self.dismissed = dismissed
    }

    private var enabledCount: Int {
        groups.filter(\.enabled).count
    }

    public var body: some View {
        MuiOverlayShell(isPresented: isPresented, accessibilityLabel: "Selective Paste", contained: contained) {
            MuiText("Selective Paste", variant: .sheetTitle)
        } content: {
            VStack(alignment: .leading, spacing: MuiTokens.spacingSm) {
                MuiText("\(enabledCount) of \(groups.count) groups will be applied.", variant: .body, color: .muted)
                ForEach(groups) { group in
                    VStack(alignment: .leading, spacing: 2) {
                        MuiCheckbox(state: group.enabled ? .checked : .unchecked, label: group.label) { toggle(group.id) }
                        if let description = group.description {
                            MuiText(description, variant: .body, color: .muted)
                                .padding(.leading, 28)
                        }
                    }
                }
            }
        } footer: {
            HStack {
                Spacer()
                MuiButton(label: "Cancel", variant: .ghost) { dismissed?() }
                MuiButton(label: "Paste", variant: .primary) { confirmPaste() }
            }
        } dismissed: {
            dismissed?()
        }
    }

    private func toggle(_ groupId: String) {
        guard let idx = groups.firstIndex(where: { $0.id == groupId }) else { return }
        let group = groups[idx]
        groups[idx] = MuiSelectivePasteGroup(id: group.id, label: group.label, description: group.description, enabled: !group.enabled)
    }

    private func confirmPaste() {
        pasteConfirmed?(groups.filter(\.enabled).map(\.id))
    }
}

#Preview("MuiSelectivePasteModal") {
    struct Demo: View {
        @State private var open = false
        @State private var groups = [
            MuiSelectivePasteGroup(id: "light", label: "Light", description: "Exposure, contrast, highlights", enabled: true),
            MuiSelectivePasteGroup(id: "color", label: "Color", description: "White balance, HSL", enabled: false),
        ]
        var body: some View {
            ZStack {
                MuiTokens.bg
                MuiButton(label: "Open Selective Paste", variant: .primary) { open = true }
                MuiSelectivePasteModal(isPresented: open, groups: $groups, dismissed: { open = false })
            }
            .frame(width: 380, height: 300)
        }
    }
    return Demo()
}
